docs/src/design/changing_the_primal.md

```diff
@@ -0,0 +1,435 @@
+which calculates eigenvalues _"either by bisection or the dqds algorithm."_.
+
+Both of these are very good algorithms.
+LAPACK has had decades of work by experts and is one of the most trusted libraries for linear algebra.
+But they are different algorithms that give different results.
+The differences in practices are around $10^{-15}$, which while very small on absolute terms
+are as far as `Float64` is concerned a very real difference.
+
+### Matrix Division
+Roughly speaking:
+`Y=A\B` is the function that finds the least-square solution to `YA ≈ B`.
+When solving such a system, the efficient way to do so is to factorize `A` into an appropriate factorized form such as `Cholesky` or `QR` etc,
+then perform the `\` operation on the factorized form.
+The pullback of `A\B` with respect to `B` is `Ȳ-> A' \ Ȳ`.
+It should be noted that that involves computing the factorization of `A'` (the adjoint of `A`).
+In this computation the factorization of the original `A` can reused.
+Doing so can give a 4x speed-up.
+
+We don't have this in ChainRules.jl yet, because Julia is missing some definitions of `adjoint` of factorizations ([JuliaLang/julia#38293](https://github.com/JuliaLang/julia/issues/38293)).
+I have been promised them for Julia v1.7 though.
+You can see what the code would look like in [PR #302](https://github.com/JuliaDiff/ChainRules.jl/pull/302).
+
+
+# Conclusion
+This document has explained why [`rrule`](@ref) is the way it is.
+In particular it has highlighted why the primal computation is able to be changed from simply calling the function.
+Futher, it has explained why `rrule` returns a closure for the pullback, rather than it being a seperate function.
+It has highlighted several places in ChainRules.jl where this has allowed us to have significantly improved performance.
+Being able to change the primal computation is practically essential for a high performance AD system.
+
+
+
+[^1]:
+    I am not just picking on Nabla randomly.
+    Many of the core developers of ChainRules worked on Nabla prior.
+    It's a good AD, but ChainRules incorporates lessons learned from working on Nabla.
+
+[^2]: which may be an explicit tape, or an implicit tape that is actually incorporated into generated code (ala Zygote)
+
+[^3]:
+    Another key reason is if the operations is a primitive that is not defined in terms of more basic operations.
+    In many languages this is the case for `sin`; where the actual implementation is in some separate `libm.so`.
+    But actually [`sin` in Julia is defined in terms of a polynomial](https://github.com/JuliaLang/julia/blob/caeaceff8af97565334f35309d812566183ec687/base/special/trig.jl).
+    It's fairly vanilla Julia code.
+    It shouldn't be too hard for an AD that only knows about basic operations like `+` and `*` to AD through it.
+    Though that will incur something that looks a lot like truncation error (in apparent violation of Griewank and Walther's 0th Rule of AD).
```

Comment:
I have no idea what this means? Can you elaborate on it? Or if not, just not make it at all?

Reply:
Yeah, it belongs in its own discussion.
I will delete

```suggestion
```